File: src/imgproc_service_client_main.cpp
#include <ros/ros.h>
#include <image_transport/image_transport.h>
#include <opencv2/highgui.hpp>
#include <cv_bridge/cv_bridge.h>
#include <meteor/SrvTutorial.h>
#include <meteor/testlibcall.h>

int main(int argc, char** argv)
{
    // test lib call
    int a = 1, b = 1;
    int c = addTwoInts(a, b);

    cv::Mat img = cv::imread("src/meteor/data/panda.jpeg");
    sensor_msgs::ImagePtr im_msg = cv_bridge::CvImage(std_msgs::Header(), "bgr8", img).toImageMsg();

    ros::init(argc, argv, "imgproc_service_client_main");
    ros::NodeHandle nh;
    ros::ServiceClient client = nh.serviceClient<meteor::SrvTutorial>("image_flip");
    meteor::SrvTutorial srv;
    srv.request.a = 1;
    srv.request.b = 2;
    srv.request.input = *im_msg;

    if(client.call(srv))
    {
        ROS_INFO("send srv, srv.Request.a and b: %ld, %ld", (long int)srv.request.a, (long int)srv.request.b);
        ROS_INFO("receive srv, srv.Response.result: %ld", (long int)srv.response.result);

        cv_bridge::CvImagePtr output_ptr = cv_bridge::toCvCopy(srv.response.output, "bgr8");
        cv::namedWindow("output");
        cv::imshow("output", output_ptr->image);
        cv::waitKey();
    }
    else
    {
        ROS_ERROR("fail to call service image_flip");
        return 1;
    }

    return 0;
}

File: scripts/imgproc_service_client_main.py
#! /usr/bin/env python

from meteor.srv import *
from sensor_msgs.msg import Image
from cv_bridge import CvBridge, CvBridgeError
import rospy
import cv2

if __name__ == "__main__":

    img = cv2.imread('../data/panda.jpeg')

    rospy.init_node('imgproc_service_client_main')
    # rospy.wait_for_service('image_flip')
    s = rospy.ServiceProxy('image_flip', SrvTutorial)

    req = SrvTutorialRequest()
    req.a = 1
    req.b = 2
    cvbridge = CvBridge()
    req.input = cvbridge.cv2_to_imgmsg(img, 'bgr8')

    try:
        res = s.call(req)
        rospy.loginfo("send srv, srv.Request.a and b: %ld, %ld", req.a, req.b)
        rospy.loginfo("receive srv, srv.Response.result: %ld", res.result)

        flip_img = cvbridge.imgmsg_to_cv2(res.output, 'bgr8')
        rospy.loginfo("response output image rows: %d, cols: %d", flip_img.shape[0], flip_img.shape[1])

        # Chris TODO: imshow of python-opencv often leads to unexpected errors
        cv2.namedWindow('output')
        cv2.imshow('output', flip_img)
        cv2.waitKey()
        cv2.destroyAllWindows()

        s.close()

    except rospy.ServiceException as exc:
        print("Service did not process request: " + str(exc))
